Implement FetchX509Bundles and support unregistered retrievals of x509 bundles

srwaggon opened the pull request:
<!--
1. If this is your first PR, please read our contributor guidelines
https://github.com/spiffe/spiffe/blob/master/CONTRIBUTING.md
https://github.com/spiffe/spire/blob/master/CONTRIBUTING.md

2. Please remember to include a DCO on every commit (`git commit -s`)
https://github.com/apps/dco

-->

**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [x] Documentation updated?
- [x] Include DCO

**Affected functionality**
<!-- Please provide a description of the affected functionality -->
Implement the FetchX509Bundles rpc to allow workload api clients to easily get updates for the x509 bundles.

**Description of change**
<!-- Please provide a description of the change -->
We introduced a new configuration flag. This flag is:
`-allowUnregisteredClientsToFetchX509Bundles` as a command flag or
`allow_unregistered_clients_to_fetch_x509_bundles` in the configuration file.

See test/fixture/config/agent_good.conf for an example conf file usage.

This configuration will allow clients that do not have a workload registration entry to retrieve x509 bundles from the spire-agent using the WorkloadAPI.

These unregistered clients will only receive the bundles for the trust domain of the agent.

**Which issue this PR fixes**
<!-- optional. `fixes #<issue number>` format will close an issue when this PR is merged -->
fixes #2089 


There are still a few things we need to do for this changeset. We are hoping to get some feedback on the naming while we polish it up. 

The config/flag names are `allow_unregistered_clients_to_fetch_x509_bundles`/`allowUnregisteredClientsToFetchX509Bundles`. Do these names sound like the user experience that SPIRE is looking for? Do you have a preferred name for this property?

There is a temporary commit pointing to a forked go-spiffe in the go.mod. We are unable to use the upstream until the [go-spiffe RPC PR](https://github.com/spiffe/go-spiffe/pull/159) is merged.

- with @adobley

amartinezfayo commented:
Thank you @srwaggon for this contribution!

> There are still a few things we need to do for this changeset. We are hoping to get some feedback on the naming while we polish it up.
> 
> The config/flag names are allow_unregistered_clients_to_fetch_x509_bundles/allowUnregisteredClientsToFetchX509Bundles. Do these names sound like the user experience that SPIRE is looking for? Do you have a preferred name for this property?

I'm not sure if "unregistered clients" is the best way to refer to unauthenticated workloads. The term "client" may be confusing for example if it's a server that is making the request. Also, the workload may be "registered" but still not be able to retrieve an identity based on the selectors resolved. So to me, referring to "authenticated / unauthenticated" would be more precise.
I think that something along the lines of `allow_unauthenticated_verifiers` as suggested by @azdagron in #1834 is a better choice for this, that can be used for FetchX509Bundles and (eventually) FetchJWTBundles.

srwaggon commented:
(with -@adobley)

We've updated the flag name from `-allowUnregisteredClientsToFetchX509Bundles` to `-allowUnauthenticatedVerifiers` as per suggestion. 

We have also updated the documentation to reflect this new flag.
We are also still awaiting the merging of the [RPC for fetching x509 bundles](https://github.com/spiffe/go-spiffe/pull/159) so that the go.mod can be correctly updated.

adobley commented:
https://github.com/spiffe/go-spiffe/pull/159 was merged so we've updated the commits to reference the new go-spiffe v2.0.0-beta.5 tag.

This changeset should be good to go!

@srwaggon can you change this from a draft when you have the time? Thanks!

MarcosDY commented on pkg/agent/endpoints/workload/handler.go:
```diff
@@ -205,6 +206,65 @@ func (h *Handler) FetchX509SVID(_ *workload.X509SVIDRequest, stream workload.Spi
+
+func composeX509BundlesResponse(update *cache.WorkloadUpdate) *workload.X509BundlesResponse {
+	bundles := make(map[string][]byte)
+	if update.Bundle != nil {
```
based on SPEC ```The `X509BundlesResponse` response message has a mandatory `bundles` field, which MUST contain at least the trust bundle for the trust domain in which the server resides. The `crl` field is optional.```
if Bundle is nil, may we return an error?

azdagron reviewed (approved):
Thank you for this @adobley and @srwaggon! I've put this through it's paces and it looks good.

azdagron commented:
Looks like the update broke the PR. I'll push a commit to patch.

azdagron commented:
Also added the defensive check while I was in there.

MarcosDY reviewed (approved):
LGTM!

azdagron reviewed (approved):
Thanks for taking a look, @MarcosDY !
